### DISCUSSION TITLE
how to run xterm-headless without nodejs

### DISCUSSION THREAD

**Discussion by BinaryRyan** [sourcelair/xterm.js#5248] [Q&A] [answered]
I am looking for a way to run xterm headless without relying on the node environment. Is there any way to compile xterm.js into a dynamic library for use, using Rust or CPP,If using a node environment, I am not sure if its speed will affect the state of the mirror server terminal.And I also need to increase communication methods to obtain the status of the mirror terminal,it's not good.

**Comment by jerch** [comment, marked as answer]
Hmm, in theory this should be possible as xterm.js does not rely on specific parts of the node env per se. We have only one place, where we deduct the underlying platform in `src/common/Platform.ts`. As you can see there you'd get node behavior by defining a global `process` object.
To find follow-ups in the code you can search for `isNode` and check impl details at the occasions and whether you can polyfill it in your JS engine. A quick search myself revealed only one impl switch here: https://github.com/xtermjs/xterm.js/blob/e826f6b69f42a2b0c5a3a532300db58be70b50e9/src/common/TaskQueue.ts#L145

Idk if the bundler for the headless version further injects node specific shims, if so, you might be able to polyfill them as well. 

Regarding JS engine caps - things like array buffer / dataview types, Math, console, setTimeout should be available. Also property accessors are used here and there. The core gets compiled with ES2021 currently (might still work down to ES2017 as we changed that not long ago).

In summary - yes should be possible, as xterm.js internals are all quite vanilla JS, but you are on your own to polyfill missing pieces in your JS engine. A good starting point to mess with would be the headless bundle in the _@xterm/headless_ package under `/lib-headless/xterm-headless.js`.
